### ISSUE TITLE
[BUG][ServiceBus] Cannot reuse ServiceBusClient to read from 2 topics in the same namespace

### ISSUE THREAD

**Issue by dzendras** [Azure/azure-sdk-for-net#30947]
### Library name and version

Azure.Messaging.ServiceBus 7.10.0

### Describe the bug

Despite what the docs say:

> The ServiceBusClient, senders, receivers, and processors are safe to cache and use as a singleton for the lifetime of the application, which is best practice when messages are being sent or received regularly. They are responsible for efficient management of network, CPU, and memory use, working to keep usage low during periods of inactivity.

It seems that I cannot reuse a ServiceBusClient instance in order to read from 2 topics from the same namespace. The first PeekLock succeeds, while the second one always fails with UnauthorizedAccessException:

Put token failed. status-code: 401, status-description: InvalidAudience: The authorization header contains a token with a wrong audience..

It does not matter which topic I read the first. The first one can be read from in the context of the ServiceBusClient. If I create a new client and swap topics, still the first one can be read from, while the attempt to read from the second fails.

### Expected behavior

Reading topics. topic1 = A, topic2 = B
PeekLock succeded for A/Subscriptions/default
PeekLock succeded for B/Subscriptions/default
PeekLock succeded for B/Subscriptions/default
PeekLock succeded for A/Subscriptions/default
Reading topics. topic1 = B, topic2 = A
PeekLock succeded for B/Subscriptions/default
PeekLock succeded for A/Subscriptions/default
PeekLock succeded for A/Subscriptions/default
PeekLock succeded for B/Subscriptions/default

### Actual behavior

Reading topics. topic1 = A, topic2 = B
PeekLock succeded for A/Subscriptions/default
PeekLock failed for B/Subscriptions/default
PeekLock failed for B/Subscriptions/default
PeekLock succeded for A/Subscriptions/default
Reading topics. topic1 = B, topic2 = A
PeekLock succeded for B/Subscriptions/default
PeekLock failed for A/Subscriptions/default
PeekLock failed for A/Subscriptions/default
PeekLock succeded for B/Subscriptions/default

### Reproduction Steps

```
var sendReceiveConnectionString = config.GetValue<string>("sendReceiveConnectionString");
//Endpoint=sb://NAMESPACE.servicebus.windows.net/;SharedAccessKeyName=SAKN;SharedAccessKey=SAK

async Task ReadTopics(string topic1, string topic2)
{
    Console.WriteLine($"Reading topics. topic1 = {topic1}, topic2 = {topic2}");
    var serviceBusClient = new ServiceBusClient(sendReceiveConnectionString);

    const string subscriptionName = "default";

    await using var receiver1 = serviceBusClient.CreateReceiver(topic1, subscriptionName);
    await using var receiver2 = serviceBusClient.CreateReceiver(topic2, subscriptionName);

    async Task PeekMessagesAsync(ServiceBusReceiver receiver)
    {
        const int messagesCount = 1;
        try
        {
            await receiver.PeekMessagesAsync(messagesCount);
            Console.WriteLine($"PeekLock succeded for {receiver.EntityPath}");
        }
        catch (UnauthorizedAccessException)
        {
            Console.WriteLine($"PeekLock failed for {receiver.EntityPath}");
        }
    }

    foreach (var receiver in new[] { receiver1, receiver2 })
    {
        await PeekMessagesAsync(receiver);
    }
    foreach (var receiver in new[] { receiver2, receiver1 })
    {
        await PeekMessagesAsync(receiver);
    }
}

var topicA = config.GetValue<string>("topicA");
var topicB = config.GetValue<string>("topicB");

await ReadTopics(topicA, topicB);
await ReadTopics(topicB, topicA);
```

**Comment by jsquire** [comment]
Hi @dzendras.  Thank you for reaching out and we regret that you're experiencing difficulties.  I'm able to use the code that you've shared to create both senders and receivers for multiple queues/topics/subscriptions under the same namespace.

From the scenario you're describing, it seems as if you're likely initializing the client with a shared access key that doesn't have rights to your second topic.  In order to work with multiple messaging entities under the namespace, you'll need to use a namespace-level shared access policy.  Would you please be so kind as to test using the `RootManageSharedAccessKey` for the namespace?

**Comment by dzendras** [comment]
Hi @jsquire ,

Thanks for checking. If you take a closer look at my repro (ReadTopics(A, B) and then ReadTopics(B, A)), you'll see that I'm able to query both topics, but not with the same instance of ServiceBusClient. This proves that both topics can be read with the same shared access key:

PeekLock succeded for A/Subscriptions/default
PeekLock succeded for B/Subscriptions/default

**Comment by jsquire** [comment]
@dzendras:  Do you, by chance, have an entity level shared access policy with the same name as the namespace-level policy in use?  The only time that we've seen something similar to what you're describing was due to how Service Bus resolves access policies on the service, preferring the entity-level policy.  

I'm unable to reproduce your scenario, including when I set up policies with matching names.   Regardless of key configuration, so long as the namespace-level policy exists with "Listen" permissions, I get the following when running the snippet:

```text
Reading topics. topic1 = topicA, topic2 = topicB
PeekLock succeded for topicA/Subscriptions/subscr
PeekLock succeded for topicB/Subscriptions/subscr
PeekLock succeded for topicB/Subscriptions/subscr
PeekLock succeded for topicA/Subscriptions/subscr
Reading topics. topic1 = topicB, topic2 = topicA
PeekLock succeded for topicB/Subscriptions/subscr
PeekLock succeded for topicA/Subscriptions/subscr
PeekLock succeded for topicA/Subscriptions/subscr
PeekLock succeded for topicB/Subscriptions/subscr
```

### Next Steps:

Would you please set up a new namespace-level policy with Read, Write, and Manage permissions and test that against your Service Bus.   If that fails, would you be willing to create a new Service Bus namespace with topics/subscriptions that match your configuration and test against that?   If you're able to reproduce the behavior with a new namespace, can you help me understand how things are configured so that I can continue to try and reproduce?

**Comment by dzendras** [comment]
@jsquire I had found the issue you are referring too :) Unfortunately it appears that I have encountered a different issue. 
Namespace-level policy with Manage, Send, Listen claims has a different name than topic-level policy with Send, Listen claims.
However all topic-level policies have the same name. In the repro I use Primary Connection String taken from either topic's SAS policy.

I am not sure what do you mean by "test that against your service bus" - take the newly created namespace-level policy's Primary Connection String and use it in the repro?

**Comment by dzendras** [comment]
It might be worth noting that all topic-level policies not only share the same name but also the same Primary Key, hence it doesn't matter which connection string I take (topicA's or topicB's).

**Comment by jsquire** [comment]
> I am not sure what do you mean by "test that against your service bus" - take the newly created namespace-level policy's Primary Connection String and use it in the repro?

That's exactly what I'm suggesting.  😄 

I'd like to see what happens when we use the new namespace-level connection string with your code snippet against the Service Bus instance that you've seen failing.   If you're able to repro with the new connection string, it points us in the direction of something with the configuration of your Service Bus namespace that differs from mine.

If you're unable to repro, that would seem to confirm my speculation that the issue is related to how the service associates policies with connections.

> It might be worth noting that all topic-level policies not only share the same name but also the same Primary Key, hence it doesn't matter which connection string I take (topicA's or topicB's).

Oh, that's interesting and I'm willing to bet that's our root cause.  The way that Service Bus binds access policies at the entity level is that whichever is first associated is then bound to the connection for as long as it's open.

**Comment by dzendras** [comment]
Apologies for another question, but I'm not an expert with ServiceBus. Why can't I use the existing namespace-level policy in the first place? Why are you suggesting creating another one next to the existing one?
 
>     It might be worth noting that all topic-level policies not only share the same name but also the same Primary Key, hence it doesn't matter which connection string I take (topicA's or topicB's).
> 
> Oh, that's interesting and I'm willing to bet that's our root cause. The way that Service Bus binds access policies at the entity level is that whichever is first associated is then bound to the connection for as long as it's open.

Is this behavior documented anywhere? What do you mean by "Service Bus" in this context? I assume "Azure.Messaging.ServiceBus" and not the Azure service?

**Comment by jsquire** [comment]
> Apologies for another question, but I'm not an expert with ServiceBus. Why can't I use the existing namespace-level policy in the first place? Why are you suggesting creating another one next to the existing one?

Testing a theory.   I suspect your existing policies may be interfering with one another due to how the service binds to them.

> What do you mean by "Service Bus" in this context? I assume "Azure.Messaging.ServiceBus" and not the Azure service?

The opposite: "Service Bus" is intended to indicate "the Service Bus service".  The client library (Azure.Messaging.ServiceBus) simply passes along the connection string that you give it.   It has no knowledge of the access policies on the service, nor does it perform any authorization.   

> Is this behavior documented anywhere? 

To my knowledge, the service team has not yet documented this.  The issue was raised to them when discovered and they're tracking it on their backlog.

**Comment by dzendras** [comment]
I'll test with namespace-level connection strings tomorrow. I'm off for today already :)

If sharing primary keys across multiple topics is the root cause, then what are my options here?  
1. Use a namespace-level policy to send/receive messages for each topic in that namespace and re-use a single instance of ServiceBusClient?
2. Create a ServiceBusClient instance per topic and use topic-level policy?
3. Use unique Primary Keys for each topic and re-use a single instance of ServiceBusClient? It would require administration changes to existing namespaces, thus it is the least favorable from my perspective, while the previous ones require code alignments only.
4. ...?

Basically, are there any Service Bus Design guidelines related to policies/keys/connection strings?

**Comment by dzendras** [comment]
I tested the namespace-level policies:
1. Existing with Send, Listen & Manage
2. New one with Send & Listen only

Then used their primary connection strings, I got the desired output, sharing a single ServiceBusClient instance.
I presume that the best approach would be to use the namespace-level policy with Send & Listen claims for sending/reading messages and use the existing Send/Listen/Manage only for managing the namespace.

**Comment by jsquire** [comment]
### Things to know

- An entity-level policy is intended to be used when you want to restrict access/permissions to just that entity.   It cannot be used for the namespace itself or other entities.
 
- A namespace-level policy is intended to be used when you want to interact with multiple entities in that namespace.  

- The service determines the authorization context once per connection; any subsequent operations using that connection are bound to the initial authorization context.  Permissions are, of course, checked for every operation and take into account the entity involved, but the context doesn't change.   As a result, if you authorize against one specific entity, you will fail permission checks against the namespace or other entities.

- Each `ServiceBusClient` represents a single connection to the service.  Each sender/receiver/processor spawned from it shares that connection.  

- The Service Bus clients manage connections lazily; a connection is not established until an operation is requested that needs one.

### Applying knowledge to this scenario

The service binds to policies using good security practices and applies the principle of least privilege.   Since you have the same policy name, and key specified for multiple entities and the namespace, there is ambiguity on which you intend to use.   When you invoke your first service operation to peek, the connection is established, and the service finds a local policy for the requested topic.  Your connection is now using the authorization context for that topic.

When you attempt to peek from the subscription on TopicB, the connection exists and is bound to the authorization context of the policy from TopicA.  Because that policy exists only for TopicA, no permissions exist for any other entity and the permission check fails.

### Recommendations

- Avoid sharing names across policies; each should have a unique name if possible.  
 
- If you must share names across policies, do not use the same shared key value; this circumvents the security of using an entity-level policy and causes an inability to clearly specify your intent when establishing a connection.

- If you to interact with multiple entities in a namespace, use a single, long-lived `ServiceBusClient` with either an AAD identity or a namespace-level policy.  For high concurrency or extreme throughput needs, using multiple clients is may help - though each will maintain its own connection to the service.  Too many can cause network contention and performance issues.  It is recommended to start with one and then scale up if testing shows it is needed.

**Comment by jsquire** [comment]
I'm glad that we managed to get to the bottom of things.   My last summary must have crossed your test results.   I've answered your last question below.   I'm going to mark this as addressed, but please feel free to unresolve if you'd like to continue the discussion.

> Unless there are any performance considerations when using one policy for multiple topics vs using a unique policy per topic?

There are no performance considerations related to policies.   These are security scope trade-offs only.
